handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.timeout;
+
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This handler supports trivial emulation of soTimeout for clients, measure time between
+ * last <code>write()</code> method invocation and first proceed <code>channelRead()</code>
+ *
+ * NB: Timeout measurement in default implementation
+ * will be cancelled just after first potentially partial <code>read()</code> in channel, so,
+ * if you have marker of last event - it's worth to override {@link timeoutShouldBeCancelled} method,
+ * otherwise - use this handler in pair with reasonable {@link IdleStateHandler} to catch situations
+ * of extremely slow or dead reads.
+ *
+ */
+public class ReadAfterWriteTimeoutHandler extends ChannelDuplexHandler {
```

Comment:
Looks like there is no tests for *TimeoutHandler or IdleStateHandler now in netty - will figure out and prepare test for this class.